unable to compile on OSX

## Required information

**Operating system:**
Mojave

**Compiler version:**
clang-1001.0.46.4

**Observed result or behaviour:**
```
Starting >>> iceperf
--- stderr: iceoryx_introspection
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
LIB_TINFO
    linked by target "iceoryx_introspection" in directory /Users/karsten/workspace/ros2/iceoryx_ws/osx/src/iceoryx/tools/introspection

---
```

**Expected result or behaviour:**
clean compilation

**Conditions where it occurred / Performed steps:**
`colcon build` in a workspace with `iceoryx` and `rmw_iceoryx`.

--- a/iceoryx_posh/include/iceoryx_posh/roudi/iceoryx_roudi_components.hpp
+++ b/iceoryx_posh/include/iceoryx_posh/roudi/iceoryx_roudi_components.hpp
@@ -48,9 +48,9 @@ struct IceOryxRouDiComponents
             // if there is an outdated mqueue, the start of the apps will be terminated
             runtime::MqBase::cleanupOutdatedMessageQueue(MQ_ROUDI_NAME);
 
-            m_rouDiMemoryManager.createAndAnnounceMemory().on_error([](cxx::expected<RouDiMemoryManagerError> error) {
+            m_rouDiMemoryManager.createAndAnnounceMemory().or_else([](RouDiMemoryManagerError error) {
                 /// @todo errorHandler
-                LogFatal() << "Could not create SharedMemory! Error: " << static_cast<uint64_t>(error.get_error());
+                LogFatal() << "Could not create SharedMemory! Error: " << static_cast<uint64_t>(error);
                 std::terminate();
             });
             return true;

--- a/iceoryx_posh/include/iceoryx_posh/roudi/roudi_app.hpp
+++ b/iceoryx_posh/include/iceoryx_posh/roudi/roudi_app.hpp
@@ -82,7 +82,7 @@ class RouDiApp
     RouDiConfig_t m_config;
 
     posix::Semaphore m_semaphore = std::move(posix::Semaphore::create(0u)
-                                                 .on_error([] {
+                                                 .or_else([](posix::SemaphoreError&) {
                                                      std::cerr << "Unable to create the semaphore for RouDi"
                                                                << std::endl;
                                                      std::terminate();

--- a/iceoryx_posh/source/popo/building_blocks/chunk_queue_popper.cpp
+++ b/iceoryx_posh/source/popo/building_blocks/chunk_queue_popper.cpp
@@ -77,8 +77,10 @@ uint64_t ChunkQueuePopper::size() noexcept
 void ChunkQueuePopper::setCapacity(const uint64_t newCapacity) noexcept
 {
     /// @todo fix getCapacity and setCapacity issue in queues (uint32 vs uint64)
-    // this needs to be properly fixed by harmonizing the types across the functions, but currently this cast is also sufficient
-    getMembers()->m_queue.setCapacity(static_cast<decltype(MemberType_t::MAX_CAPACITY)>(newCapacity));
+    // this needs to be properly fixed by harmonizing the types across the functions, but currently this cast is also
+    // sufficient
+    getMembers()->m_queue.setCapacity(
+        static_cast<std::remove_const<decltype(MemberType_t::MAX_CAPACITY)>::type>(newCapacity));
 }
 
 uint64_t ChunkQueuePopper::getCurrentCapacity() const noexcept
@@ -100,9 +102,9 @@ void ChunkQueuePopper::clear() noexcept
         {
             // PRQA S 4117 4 # d'tor of SharedChunk will release the memory, so RAII has the side effect here
             auto chunkTupleOut = *retVal;
-            auto chunkManagement = iox::relative_ptr<mepoo::ChunkManagement>(chunkTupleOut.m_chunkOffset,
-                                                                             chunkTupleOut.m_segmentId); 
-            auto chunk = mepoo::SharedChunk(chunkManagement);                                            
+            auto chunkManagement =
+                iox::relative_ptr<mepoo::ChunkManagement>(chunkTupleOut.m_chunkOffset, chunkTupleOut.m_segmentId);
+            auto chunk = mepoo::SharedChunk(chunkManagement);
         }
         else
         {

--- a/iceoryx_posh/source/popo/building_blocks/chunk_receiver.cpp
+++ b/iceoryx_posh/source/popo/building_blocks/chunk_receiver.cpp
@@ -21,7 +21,7 @@ namespace iox
 namespace popo
 {
 ChunkReceiver::ChunkReceiver(cxx::not_null<MemberType_t* const> chunkReceiverDataPtr) noexcept
-    : ChunkQueuePopper(static_cast<ChunkQueuePopper::MemberType_t* const>(chunkReceiverDataPtr))
+    : ChunkQueuePopper(static_cast<ChunkQueuePopper::MemberType_t*>(chunkReceiverDataPtr))
 {
 }
 

--- a/iceoryx_posh/source/roudi/application/roudi_main.cpp
+++ b/iceoryx_posh/source/roudi/application/roudi_main.cpp
@@ -28,14 +28,12 @@ int main(int argc, char* argv[])
 
     iox::roudi::TomlRouDiConfigFileProvider configFileProvider(cmdLineParser);
 
-    using ParseResult = iox::cxx::expected<iox::RouDiConfig_t, iox::roudi::RouDiConfigFileParseError>;
-
     iox::RouDiConfig_t roudiConfig =
         configFileProvider.parse()
-            .on_error([](ParseResult& parseResult) {
+            .or_else([](iox::roudi::RouDiConfigFileParseError& parseResult) {
                 iox::LogFatal() << "Couldn't parse config file. Error: "
                                 << iox::cxx::convertEnumToString(iox::roudi::ROUDI_CONFIG_FILE_PARSE_ERROR_STRINGS,
-                                                                 parseResult.get_error());
+                                                                 parseResult);
                 std::terminate();
             })
             .get_value();

--- a/iceoryx_posh/source/runtime/message_queue_interface.cpp
+++ b/iceoryx_posh/source/runtime/message_queue_interface.cpp
@@ -87,9 +87,7 @@ bool MqBase::receive(MqMessage& answer) const noexcept
 bool MqBase::timedReceive(const units::Duration timeout, MqMessage& answer) const noexcept
 {
     return !m_mq.timedReceive(timeout)
-                .on_success([&answer](cxx::expected<std::string, posix::IpcChannelError>& message) {
-                    MqBase::setMessageFromString(message.get_value().c_str(), answer);
-                })
+                .and_then([&answer](std::string& message) { MqBase::setMessageFromString(message.c_str(), answer); })
                 .has_error()
            && answer.isValid();
 }
@@ -114,15 +112,15 @@ bool MqBase::send(const MqMessage& msg) const noexcept
         return false;
     }
 
-    auto logLengthError = [&msg](cxx::expected<posix::IpcChannelError>& error) {
-        if (error.get_error() == posix::IpcChannelError::MESSAGE_TOO_LONG)
+    auto logLengthError = [&msg](posix::IpcChannelError& error) {
+        if (error == posix::IpcChannelError::MESSAGE_TOO_LONG)
         {
             const size_t messageSize =
                 static_cast<size_t>(msg.getMessage().size()) + posix::MessageQueue::NULL_TERMINATOR_SIZE;
             LogError() << "msg size of " << messageSize << "bigger than configured max message size";
         }
     };
-    return !m_mq.send(msg.getMessage()).on_error(logLengthError).has_error();
+    return !m_mq.send(msg.getMessage()).or_else(logLengthError).has_error();
 }
 
 bool MqBase::timedSend(const MqMessage& msg, units::Duration timeout) const noexcept
@@ -134,15 +132,15 @@ bool MqBase::timedSend(const MqMessage& msg, units::Duration timeout) const noex
         return false;
     }
 
-    auto logLengthError = [&msg](cxx::expected<posix::IpcChannelError>& error) {
-        if (error.get_error() == posix::IpcChannelError::MESSAGE_TOO_LONG)
+    auto logLengthError = [&msg](posix::IpcChannelError& error) {
+        if (error == posix::IpcChannelError::MESSAGE_TOO_LONG)
         {
             const size_t messageSize =
                 static_cast<size_t>(msg.getMessage().size()) + posix::MessageQueue::NULL_TERMINATOR_SIZE;
             LogError() << "msg size of " << messageSize << "bigger than configured max message size";
         }
     };
-    return !m_mq.timedSend(msg.getMessage(), timeout).on_error(logLengthError).has_error();
+    return !m_mq.timedSend(msg.getMessage(), timeout).or_else(logLengthError).has_error();
 }
 
 const std::string& MqBase::getInterfaceName() const noexcept
@@ -162,7 +160,7 @@ bool MqBase::openMessageQueue(const posix::IpcChannelSide channelSide) noexcept
     m_channelSide = channelSide;
     IpcChannelType::create(
         m_interfaceName, posix::IpcChannelMode::BLOCKING, m_channelSide, m_maxMessageSize, m_maxMessages)
-        .on_success([this](cxx::expected<IpcChannelType, posix::IpcChannelError>& mq) { this->m_mq = std::move(*mq); });
+        .and_then([this](IpcChannelType& mq) { this->m_mq = std::move(mq); });
 
     return m_mq.isInitialized();
 }

--- a/iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp
+++ b/iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp
@@ -32,7 +32,7 @@ using ::testing::Return;
 
 struct DummySample
 {
-    uint64_t dummy{42};
+    uint64_t m_dummy{42};
 };
 
 static constexpr uint32_t NUM_CHUNKS_IN_POOL = 3 * iox::MAX_RECEIVER_QUEUE_CAPACITY;
@@ -72,14 +72,14 @@ class ChunkBuildingBlocks_IntegrationTest : public Test
         for (size_t i = 0; i < ITERATIONS; i++)
         {
             m_chunkSender.allocate(sizeof(DummySample))
-                .on_success([&](iox::mepoo::ChunkHeader* chunkHeader) {
+                .and_then([&](iox::mepoo::ChunkHeader* chunkHeader) {
                     auto sample = chunkHeader->payload();
                     new (sample) DummySample();
-                    static_cast<DummySample*>(sample)->dummy = i;
+                    static_cast<DummySample*>(sample)->m_dummy = i;
                     m_chunkSender.send(chunkHeader);
                     m_sendCounter++;
                 })
-                .on_error([]() {
+                .or_else([](AllocationError) {
                     // Errors shall never occur
                     FAIL();
                 });
@@ -106,7 +106,7 @@ class ChunkBuildingBlocks_IntegrationTest : public Test
                 .and_then([&](SharedChunk& chunk) {
                     auto dummySample = *reinterpret_cast<DummySample*>(chunk.getPayload());
                     // Check if monotonically increasing
-                    EXPECT_THAT(dummySample.dummy, Eq(forwardCounter));
+                    EXPECT_THAT(dummySample.m_dummy, Eq(forwardCounter));
                     m_chunkDistributor.deliverToAllStoredQueues(chunk);
                     forwardCounter++;
                     newChunkReceivedInLastIteration = true;
@@ -141,13 +141,13 @@ class ChunkBuildingBlocks_IntegrationTest : public Test
             ASSERT_FALSE(m_chunkReceiver.hasOverflown());
 
             m_chunkReceiver.get()
-                .on_success([&](iox::cxx::optional<const iox::mepoo::ChunkHeader*>& maybeChunkHeader) {
+                .and_then([&](iox::cxx::optional<const iox::mepoo::ChunkHeader*>& maybeChunkHeader) {
                     if (maybeChunkHeader.has_value())
                     {
                         auto chunkHeader = maybeChunkHeader.value();
                         auto dummySample = *reinterpret_cast<DummySample*>(chunkHeader->payload());
                         // Check if monotonically increasing
-                        EXPECT_THAT(dummySample.dummy, Eq(m_receiveCounter));
+                        EXPECT_THAT(dummySample.m_dummy, Eq(m_receiveCounter));
                         m_receiveCounter++;
                         m_chunkReceiver.release(chunkHeader);
                         newChunkReceivedInLastIteration = true;
@@ -164,7 +164,7 @@ class ChunkBuildingBlocks_IntegrationTest : public Test
                         }
                     }
                 })
-                .on_error([]() {
+                .or_else([](ChunkReceiveError) {
                     // Errors shall never occur
                     FAIL();
                 });

--- a/iceoryx_utils/include/iceoryx_utils/cxx/expected.hpp
+++ b/iceoryx_utils/include/iceoryx_utils/cxx/expected.hpp
@@ -15,6 +15,7 @@
 #define IOX_UTILS_CXX_EXPECTED_HPP
 
 #include "iceoryx_utils/cxx/function_ref.hpp"
+#include "iceoryx_utils/cxx/optional.hpp"
 #include "iceoryx_utils/cxx/variant.hpp"
 
 #include <utility>
@@ -117,10 +118,10 @@ class expected;
 ///     }
 ///
 ///     cxx::expected<float> errorOnlyMethod() {
-///         return callMe().on_error([]{
+///         return callMe().or_else([]{
 ///             std::cerr << "Error Occured\n";
 ///             /// perform some action
-///         }).on_success([](cxx::expected<int, float> & result){
+///         }).and_then([](cxx::expected<int, float> & result){
 ///             std::cout << "Success, got " << result.get_value() << std::endl;
 ///             /// perform some action
 ///         });
@@ -243,7 +244,7 @@ class expected<ErrorType>
     ///         std::cout << "error occured : " << result.get_error() << std::endl;
     ///     })
     /// @endcode
-    const expected& on_error(const cxx::function_ref<void(expected&)>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_error(const cxx::function_ref<void(expected&)>& callable) const noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
@@ -254,29 +255,31 @@ class expected<ErrorType>
     ///         std::cout << "error occured : " << result.get_error() << std::endl;
     ///     })
     /// @endcode
-    expected& on_error(const cxx::function_ref<void(expected&)>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_error(const cxx::function_ref<void(expected&)>& callable) noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called and
     ///         a reference to the ErrorType is given as an argument to the callable
     /// @param[in] callable callable which will be called if the expected contains an error
     /// @return const reference to the expected itself
     /// @code
-    ///     someExpected.on_error([](float& error){
+    ///     someExpected.or_else([](float& error){
     ///         std::cout << "error occured : " << error << std::endl;
     ///     })
     /// @endcode
-    const expected& on_error(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_error(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept;
+    const expected& or_else(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called and
     ///         a reference to the ErrorType is given as an argument to the callable
     /// @param[in] callable callable which will be called if the expected contains an error
     /// @return const reference to the expected itself
     /// @code
-    ///     someExpected.on_error([](float& error){
+    ///     someExpected.or_else([](float& error){
     ///         std::cout << "error occured : " << error << std::endl;
     ///     })
     /// @endcode
-    expected& on_error(const cxx::function_ref<void(ErrorType&)>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_error(const cxx::function_ref<void(ErrorType&)>& callable) noexcept;
+    expected& or_else(const cxx::function_ref<void(ErrorType&)>& callable) noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called
     /// @param[in] callable callable which will be called if the expected contains an error
@@ -286,7 +289,7 @@ class expected<ErrorType>
     ///         std::cout << "error occured " << std::endl;
     ///     })
     /// @endcode
-    const expected& on_error(const cxx::function_ref<void()>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_error(const cxx::function_ref<void()>& callable) const noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called
     /// @param[in] callable callable which will be called if the expected contains an error
@@ -296,7 +299,7 @@ class expected<ErrorType>
     ///         std::cout << "error occured " << std::endl;
     ///     })
     /// @endcode
-    expected& on_error(const cxx::function_ref<void()>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_error(const cxx::function_ref<void()>& callable) noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
@@ -307,7 +310,7 @@ class expected<ErrorType>
     ///         std::cout << "we are successful" << std::endl;
     ///     })
     /// @endcode
-    const expected& on_success(const cxx::function_ref<void(expected&)>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_success(const cxx::function_ref<void(expected&)>& callable) const noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
@@ -318,29 +321,31 @@ class expected<ErrorType>
     ///         std::cout << "we are successful" << std::endl;
     ///     })
     /// @endcode
-    expected& on_success(const cxx::function_ref<void(expected&)>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_success(const cxx::function_ref<void(expected&)>& callable) noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
     /// @param[in] callable callable which will be called if the expected contains a success value
     /// @return const reference to the expected itself
     /// @code
-    ///     someExpected.on_success([]{
+    ///     someExpected.and_then([]{
     ///         std::cout << "we are successful!" << std::endl;
     ///     })
     /// @endcode
-    const expected& on_success(const cxx::function_ref<void()>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_success(const cxx::function_ref<void()>& callable) const noexcept;
+    const expected& and_then(const cxx::function_ref<void()>& callable) const noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
     /// @param[in] callable callable which will be called if the expected contains a success value
     /// @return const reference to the expected itself
     /// @code
-    ///     someExpected.on_success([]{
+    ///     someExpected.and_then([]{
     ///         std::cout << "we are successful!" << std::endl;
     ///     })
     /// @endcode
-    expected& on_success(const cxx::function_ref<void()>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_success(const cxx::function_ref<void()>& callable) noexcept;
+    expected& and_then(const cxx::function_ref<void()>& callable) noexcept;
 
   private:
     expected(variant<ErrorType>&& store, const bool hasError) noexcept;
@@ -550,7 +555,7 @@ class expected<ValueType, ErrorType>
     ///         std::cout << "error occured : " << result.get_error() << std::endl;
     ///     })
     /// @endcode
-    const expected& on_error(const cxx::function_ref<void(expected&)>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_error(const cxx::function_ref<void(expected&)>& callable) const noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
@@ -561,29 +566,31 @@ class expected<ValueType, ErrorType>
     ///         std::cout << "error occured : " << result.get_error() << std::endl;
     ///     })
     /// @endcode
-    expected& on_error(const cxx::function_ref<void(expected&)>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_error(const cxx::function_ref<void(expected&)>& callable) noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called and
     ///         a reference to the ErrorType is given as an argument to the callable
     /// @param[in] callable callable which will be called if the expected contains an error
     /// @return const reference to the expected itself
     /// @code
-    ///     someExpected.on_error([](float& result){
+    ///     someExpected.or_else([](float& result){
     ///         std::cout << "error occured : " << error << std::endl;
     ///     })
     /// @endcode
-    const expected& on_error(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_error(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept;
+    const expected& or_else(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called and
     ///         a reference to the ErrorType is given as an argument to the callable
     /// @param[in] callable callable which will be called if the expected contains an error
     /// @return reference to the expected itself
     /// @code
-    ///     someExpected.on_error([](float& error){
+    ///     someExpected.or_else([](float& error){
     ///         std::cout << "error occured : " << error << std::endl;
     ///     })
     /// @endcode
-    expected& on_error(const cxx::function_ref<void(ErrorType&)>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_error(const cxx::function_ref<void(ErrorType&)>& callable) noexcept;
+    expected& or_else(const cxx::function_ref<void(ErrorType&)>& callable) noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called
     /// @param[in] callable callable which will be called if the expected contains an error
@@ -593,7 +600,7 @@ class expected<ValueType, ErrorType>
     ///         std::cout << "error occured " << std::endl;
     ///     })
     /// @endcode
-    const expected& on_error(const cxx::function_ref<void()>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_error(const cxx::function_ref<void()>& callable) const noexcept;
 
     /// @brief  if the expected does contain an error the given callable is called
     /// @param[in] callable callable which will be called if the expected contains an error
@@ -603,7 +610,7 @@ class expected<ValueType, ErrorType>
     ///         std::cout << "error occured " << std::endl;
     ///     })
     /// @endcode
-    expected& on_error(const cxx::function_ref<void()>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_error(const cxx::function_ref<void()>& callable) noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
@@ -614,7 +621,7 @@ class expected<ValueType, ErrorType>
     ///         std::cout << "we have a result : " << result.get_value() << std::endl;
     ///     })
     /// @endcode
-    const expected& on_success(const cxx::function_ref<void(expected&)>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_success(const cxx::function_ref<void(expected&)>& callable) const noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
@@ -625,29 +632,31 @@ class expected<ValueType, ErrorType>
     ///         std::cout << "we have a result : " << result.get_value() << std::endl;
     ///     })
     /// @endcode
-    expected& on_success(const cxx::function_ref<void(expected&)>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_success(const cxx::function_ref<void(expected&)>& callable) noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the result is given as an argument to the callable
     /// @param[in] callable callable which will be called if the expected contains a success value
     /// @return const reference to the expected
     /// @code
-    ///     someExpected.on_success([](int& result){
+    ///     someExpected.and_then([](int& result){
     ///         std::cout << "we have a result : " << result << std::endl;
     ///     })
     /// @endcode
-    const expected& on_success(const cxx::function_ref<void(ValueType&)>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_success(const cxx::function_ref<void(ValueType&)>& callable) const noexcept;
+    const expected& and_then(const cxx::function_ref<void(ValueType&)>& callable) const noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the result is given as an argument to the callable
     /// @param[in] callable callable which will be called if the expected contains a success value
     /// @return reference to the expected
     /// @code
-    ///     someExpected.on_success([](int& result){
+    ///     someExpected.and_then([](int& result){
     ///         std::cout << "we have a result : " << result << std::endl;
     ///     })
     /// @endcode
-    expected& on_success(const cxx::function_ref<void(ValueType&)>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_success(const cxx::function_ref<void(ValueType&)>& callable) noexcept;
+    expected& and_then(const cxx::function_ref<void(ValueType&)>& callable) noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
@@ -658,7 +667,7 @@ class expected<ValueType, ErrorType>
     ///         std::cout << "we are successful!" << std::endl;
     ///     })
     /// @endcode
-    const expected& on_success(const cxx::function_ref<void()>& callable) const noexcept;
+    [[gnu::deprecated]] const expected& on_success(const cxx::function_ref<void()>& callable) const noexcept;
 
     /// @brief  if the expected does contain a success value the given callable is called and
     ///         a reference to the expected is given as an argument to the callable
@@ -669,7 +678,9 @@ class expected<ValueType, ErrorType>
     ///         std::cout << "we are successful!" << std::endl;
     ///     })
     /// @endcode
-    expected& on_success(const cxx::function_ref<void()>& callable) noexcept;
+    [[gnu::deprecated]] expected& on_success(const cxx::function_ref<void()>& callable) noexcept;
+
+    optional<ValueType> to_optional() const noexcept;
 
   private:
     expected(variant<ValueType, ErrorType>&& f_store, const bool hasError) noexcept;

--- a/iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl
+++ b/iceoryx_utils/include/iceoryx_utils/internal/cxx/expected.inl
@@ -131,13 +131,13 @@ inline const ErrorType&& expected<ValueType, ErrorType>::get_error() const&& noe
 }
 
 template <typename ValueType, typename ErrorType>
-    inline ErrorType&& expected<ValueType, ErrorType>::get_error() && noexcept
+inline ErrorType&& expected<ValueType, ErrorType>::get_error() && noexcept
 {
     return std::move(*m_store.template get_at_index<1>());
 }
 
 template <typename ValueType, typename ErrorType>
-    inline ErrorType& expected<ValueType, ErrorType>::get_error() & noexcept
+inline ErrorType& expected<ValueType, ErrorType>::get_error() & noexcept
 {
     return *m_store.template get_at_index<1>();
 }
@@ -149,7 +149,7 @@ inline const ErrorType& expected<ValueType, ErrorType>::get_error() const& noexc
 }
 
 template <typename ValueType, typename ErrorType>
-    inline ValueType&& expected<ValueType, ErrorType>::get_value() && noexcept
+inline ValueType&& expected<ValueType, ErrorType>::get_value() && noexcept
 {
     return std::move(*m_store.template get_at_index<0>());
 }
@@ -167,7 +167,7 @@ inline const ValueType&& expected<ValueType, ErrorType>::get_value() const&& noe
 }
 
 template <typename ValueType, typename ErrorType>
-    inline ValueType& expected<ValueType, ErrorType>::get_value() & noexcept
+inline ValueType& expected<ValueType, ErrorType>::get_value() & noexcept
 {
     return *m_store.template get_at_index<0>();
 }
@@ -253,6 +253,25 @@ expected<ValueType, ErrorType>::on_error(const cxx::function_ref<void(ErrorType&
 
 template <typename ValueType, typename ErrorType>
 inline expected<ValueType, ErrorType>&
+expected<ValueType, ErrorType>::or_else(const cxx::function_ref<void(ErrorType&)>& callable) noexcept
+{
+    if (this->has_error())
+    {
+        callable(get_error());
+    }
+
+    return *this;
+}
+
+template <typename ValueType, typename ErrorType>
+inline const expected<ValueType, ErrorType>&
+expected<ValueType, ErrorType>::or_else(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept
+{
+    return const_cast<expected*>(this)->or_else(callable);
+}
+
+template <typename ValueType, typename ErrorType>
+inline expected<ValueType, ErrorType>&
 expected<ValueType, ErrorType>::on_error(const cxx::function_ref<void()>& callable) noexcept
 {
     if (this->has_error())
@@ -309,6 +328,25 @@ expected<ValueType, ErrorType>::on_success(const cxx::function_ref<void(ValueTyp
 }
 
 template <typename ValueType, typename ErrorType>
+inline const expected<ValueType, ErrorType>&
+expected<ValueType, ErrorType>::and_then(const cxx::function_ref<void(ValueType&)>& callable) const noexcept
+{
+    return const_cast<expected*>(this)->and_then(callable);
+}
+
+template <typename ValueType, typename ErrorType>
+inline expected<ValueType, ErrorType>&
+expected<ValueType, ErrorType>::and_then(const cxx::function_ref<void(ValueType&)>& callable) noexcept
+{
+    if (!this->has_error())
+    {
+        callable(get_value());
+    }
+
+    return *this;
+}
+
+template <typename ValueType, typename ErrorType>
 inline expected<ValueType, ErrorType>&
 expected<ValueType, ErrorType>::on_success(const cxx::function_ref<void()>& callable) noexcept
 {
@@ -343,6 +381,13 @@ inline expected<ValueType, ErrorType>::operator expected<T>() const noexcept
 {
     return const_cast<expected*>(this)->operator expected<T>();
 }
+
+template <typename ValueType, typename ErrorType>
+inline optional<ValueType> expected<ValueType, ErrorType>::to_optional() const noexcept
+{
+    return (!this->has_error()) ? optional<ValueType>(static_cast<ValueType>(this->get_value())) : nullopt;
+}
+
 
 // expected<ErrorType>
 
@@ -457,7 +502,7 @@ inline bool expected<ErrorType>::has_error() const noexcept
 }
 
 template <typename ErrorType>
-    inline ErrorType&& expected<ErrorType>::get_error() && noexcept
+inline ErrorType&& expected<ErrorType>::get_error() && noexcept
 {
     return std::move(*m_store.template get_at_index<0>());
 }
@@ -475,7 +520,7 @@ inline const ErrorType&& expected<ErrorType>::get_error() const&& noexcept
 }
 
 template <typename ErrorType>
-    inline ErrorType& expected<ErrorType>::get_error() & noexcept
+inline ErrorType& expected<ErrorType>::get_error() & noexcept
 {
     return *m_store.template get_at_index<0>();
 }
@@ -492,8 +537,8 @@ inline expected<ErrorType>& expected<ErrorType>::on_error(const cxx::function_re
 }
 
 template <typename ErrorType>
-inline const expected<ErrorType>& expected<ErrorType>::on_error(const cxx::function_ref<void(expected&)>& callable) const
-    noexcept
+inline const expected<ErrorType>&
+expected<ErrorType>::on_error(const cxx::function_ref<void(expected&)>& callable) const noexcept
 {
     return const_cast<expected*>(this)->on_error(callable);
 }
@@ -510,10 +555,28 @@ inline expected<ErrorType>& expected<ErrorType>::on_error(const cxx::function_re
 }
 
 template <typename ErrorType>
-inline const expected<ErrorType>& expected<ErrorType>::on_error(const cxx::function_ref<void(ErrorType&)>& callable) const
-    noexcept
+inline const expected<ErrorType>&
+expected<ErrorType>::on_error(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept
 {
     return const_cast<expected*>(this)->on_error(callable);
+}
+
+template <typename ErrorType>
+inline expected<ErrorType>& expected<ErrorType>::or_else(const cxx::function_ref<void(ErrorType&)>& callable) noexcept
+{
+    if (this->has_error())
+    {
+        callable(get_error());
+    }
+
+    return *this;
+}
+
+template <typename ErrorType>
+inline const expected<ErrorType>&
+expected<ErrorType>::or_else(const cxx::function_ref<void(ErrorType&)>& callable) const noexcept
+{
+    return const_cast<expected*>(this)->or_else(callable);
 }
 
 template <typename ErrorType>
@@ -528,7 +591,8 @@ inline expected<ErrorType>& expected<ErrorType>::on_error(const cxx::function_re
 }
 
 template <typename ErrorType>
-inline const expected<ErrorType>& expected<ErrorType>::on_error(const cxx::function_ref<void()>& callable) const noexcept
+inline const expected<ErrorType>&
+expected<ErrorType>::on_error(const cxx::function_ref<void()>& callable) const noexcept
 {
     return const_cast<expected*>(this)->on_error(callable);
 }
@@ -545,8 +609,8 @@ inline expected<ErrorType>& expected<ErrorType>::on_success(const cxx::function_
 }
 
 template <typename ErrorType>
-inline const expected<ErrorType>& expected<ErrorType>::on_success(const cxx::function_ref<void(expected&)>& callable) const
-    noexcept
+inline const expected<ErrorType>&
+expected<ErrorType>::on_success(const cxx::function_ref<void(expected&)>& callable) const noexcept
 {
     return const_cast<expected*>(this)->on_success(callable);
 }
@@ -563,9 +627,28 @@ inline expected<ErrorType>& expected<ErrorType>::on_success(const cxx::function_
 }
 
 template <typename ErrorType>
-inline const expected<ErrorType>& expected<ErrorType>::on_success(const cxx::function_ref<void()>& callable) const noexcept
+inline const expected<ErrorType>&
+expected<ErrorType>::on_success(const cxx::function_ref<void()>& callable) const noexcept
 {
     return const_cast<expected*>(this)->on_success(callable);
+}
+
+template <typename ErrorType>
+inline expected<ErrorType>& expected<ErrorType>::and_then(const cxx::function_ref<void()>& callable) noexcept
+{
+    if (!this->has_error())
+    {
+        callable();
+    }
+
+    return *this;
+}
+
+template <typename ErrorType>
+inline const expected<ErrorType>&
+expected<ErrorType>::and_then(const cxx::function_ref<void()>& callable) const noexcept
+{
+    return const_cast<expected*>(this)->and_then(callable);
 }
 } // namespace cxx
 } // namespace iox

--- a/iceoryx_utils/include/iceoryx_utils/internal/cxx/optional.inl
+++ b/iceoryx_utils/include/iceoryx_utils/internal/cxx/optional.inl
@@ -111,8 +111,8 @@ inline optional<T>::~optional() noexcept
 
 template <typename T>
 template <typename U>
-inline typename std::enable_if<!std::is_same<U, optional<T>&>::value, optional<T>>::type& optional<T>::
-operator=(U&& newValue) noexcept
+inline typename std::enable_if<!std::is_same<U, optional<T>&>::value, optional<T>>::type&
+optional<T>::operator=(U&& newValue) noexcept
 {
     if (m_hasValue)
     {
@@ -220,7 +220,7 @@ inline void optional<T>::reset() noexcept
 }
 
 template <typename T>
-    inline T& optional<T>::value() & noexcept
+inline T& optional<T>::value() & noexcept
 {
     auto data = (has_value()) ? static_cast<T*>(static_cast<void*>(m_data)) : nullptr;
     return *data;
@@ -234,7 +234,7 @@ inline const T& optional<T>::value() const& noexcept
 }
 
 template <typename T>
-    inline T&& optional<T>::value() && noexcept
+inline T&& optional<T>::value() && noexcept
 {
     auto data = (has_value()) ? static_cast<T*>(static_cast<void*>(m_data)) : nullptr;
     return std::move(*data);

--- a/iceoryx_utils/source/posix_wrapper/shared_memory_object/allocator.cpp
+++ b/iceoryx_utils/source/posix_wrapper/shared_memory_object/allocator.cpp
@@ -24,7 +24,7 @@ namespace posix
 {
 constexpr uint64_t Allocator::MEMORY_ALIGNMENT;
 Allocator::Allocator(void* const f_startAddress, const uint64_t f_length) noexcept
-    : m_startAddress(static_cast<byte_t* const>(f_startAddress))
+    : m_startAddress(static_cast<byte_t*>(f_startAddress))
     , m_length(f_length)
 {
     /// @todo memset to set memory and to avoid the usage of unavailable memory

--- a/iceoryx_utils/test/moduletests/test_cxx_expected.cpp
+++ b/iceoryx_utils/test/moduletests/test_cxx_expected.cpp
@@ -12,8 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "test.hpp"
 #include "iceoryx_utils/cxx/expected.hpp"
+#include "test.hpp"
 
 using namespace ::testing;
 using namespace iox::cxx;
@@ -24,24 +24,24 @@ class expected_test : public Test
     struct Test
     {
         Test(int a, int b)
-            : a(a)
-            , b(b)
+            : m_a(a)
+            , m_b(b)
         {
         }
 
-        int Gimme()
+        int gimme()
         {
-            return a + b;
+            return m_a + m_b;
         }
 
-        int ConstGimme() const
+        int constGimme() const
         {
-            return a + b;
+            return m_a + m_b;
         }
 
 
-        int a;
-        int b;
+        int m_a;
+        int m_b;
     };
 };
 
@@ -63,14 +63,14 @@ TEST_F(expected_test, CreateValue)
 {
     auto sut = expected<Test, int>::create_value(12, 222);
     ASSERT_THAT(sut.has_error(), Eq(false));
-    EXPECT_THAT(sut.get_value().a, Eq(12));
+    EXPECT_THAT(sut.get_value().m_a, Eq(12));
 }
 
 TEST_F(expected_test, CreateError)
 {
     auto sut = expected<int, Test>::create_error(313, 212);
     ASSERT_THAT(sut.has_error(), Eq(true));
-    EXPECT_THAT(sut.get_error().b, Eq(212));
+    EXPECT_THAT(sut.get_error().m_b, Eq(212));
 }
 
 TEST_F(expected_test, GetValueOrWithError)
@@ -101,14 +101,14 @@ TEST_F(expected_test, ArrowOperator)
 {
     auto sut = expected<Test, float>::create_value(55, 81);
     ASSERT_THAT(sut.has_error(), Eq(false));
-    EXPECT_THAT(sut->Gimme(), Eq(136));
+    EXPECT_THAT(sut->gimme(), Eq(136));
 }
 
 TEST_F(expected_test, ConstArrowOperator)
 {
     const expected<Test, float> sut(success<Test>(Test(55, 81)));
     ASSERT_THAT(sut.has_error(), Eq(false));
-    EXPECT_THAT(sut->ConstGimme(), Eq(136));
+    EXPECT_THAT(sut->constGimme(), Eq(136));
 }
 
 TEST_F(expected_test, Dereferencing)
@@ -165,145 +165,47 @@ TEST_F(expected_test, CreateFromErrorType)
     EXPECT_THAT(sut.get_error(), Eq(112.1f));
 }
 
-TEST_F(expected_test, OnErrorWhenHavingAnErrorWithResult)
+TEST_F(expected_test, OrElseWhenHavingAnErrorWithResult)
 {
     expected<int, float> sut{error<float>(112.1f)};
     float a = 0.2f;
-    sut.on_error([&](expected<int, float>& r) { a = r.get_error(); }).on_success([&](expected<int, float>&) {
-        a = 2.0f;
-    });
+    sut.and_then([&](int&) { a = 2.0f; }).or_else([&](float& r) { a = r; });
 
     EXPECT_THAT(a, Eq(112.1f));
 }
 
-TEST_F(expected_test, ConstOnErrorWhenHavingAnErrorWithResult)
+TEST_F(expected_test, ConstOrElseWhenHavingAnErrorWithResult)
 {
     const expected<int, float> sut{error<float>(12.1f)};
     float a = 7.1f;
-    sut.on_error([&](expected<int, float>& r) { a = r.get_error(); }).on_success([&](expected<int, float>&) {
-        a = 91.f;
-    });
+    sut.and_then([&](int&) { a = 91.f; }).or_else([&](float& r) { a = r; });
 
     EXPECT_THAT(a, Eq(12.1f));
 }
 
-TEST_F(expected_test, OnErrorWhenHavingAnErrorWithResultErrorType)
-{
-    expected<int, float> sut{error<float>(112.1f)};
-    float a = 0.2f;
-    sut.on_error([&](float& r) { a = r; }).on_success([&](expected<int, float>&) {
-        a = 2.0f;
-    });
-
-    EXPECT_THAT(a, Eq(112.1f));
-}
-
-TEST_F(expected_test, ConstOnErrorWhenHavingAnErrorWithResultErrorType)
-{
-    const expected<int, float> sut{error<float>(12.1f)};
-    float a = 7.1f;
-    sut.on_error([&](float& r) { a = r; }).on_success([&](expected<int, float>&) {
-        a = 91.f;
-    });
-
-    EXPECT_THAT(a, Eq(12.1f));
-}
-
-TEST_F(expected_test, ErrorTypeOnlyOnErrorWhenHavingAnErrorWithResult)
+TEST_F(expected_test, ErrorTypeOnlyOrElseWhenHavingAnErrorWithResultErrorType)
 {
     expected<float> sut{error<float>(7112.1f)};
     float a = 70.2f;
-    sut.on_error([&](expected<float>& r) { a = r.get_error(); }).on_success([&](expected<float>&) { a = 2.0f; });
+    sut.and_then([&]() { a = 2.0f; }).or_else([&](float& r) { a = r; });
 
     EXPECT_THAT(a, Eq(7112.1f));
 }
 
-TEST_F(expected_test, ErrorTypeOnlyConstOnErrorWhenHavingAnErrorWithResult)
+TEST_F(expected_test, ErrorTypeOnlyConstOrElseWhenHavingAnErrorWithResultErrorType)
 {
     const expected<float> sut{error<float>(612.1f)};
     float a = 67.1f;
-    sut.on_error([&](expected<float>& r) { a = r.get_error(); }).on_success([&](expected<float>&) { a = 91.f; });
+    sut.and_then([&]() { a = 91.f; }).or_else([&](float& r) { a = r; });
 
     EXPECT_THAT(a, Eq(612.1f));
 }
 
-TEST_F(expected_test, ErrorTypeOnlyOnErrorWhenHavingAnErrorWithResultErrorType)
-{
-    expected<float> sut{error<float>(7112.1f)};
-    float a = 70.2f;
-    sut.on_error([&](float& r) { a = r; }).on_success([&](expected<float>&) { a = 2.0f; });
-
-    EXPECT_THAT(a, Eq(7112.1f));
-}
-
-TEST_F(expected_test, ErrorTypeOnlyConstOnErrorWhenHavingAnErrorWithResultErrorType)
-{
-    const expected<float> sut{error<float>(612.1f)};
-    float a = 67.1f;
-    sut.on_error([&](float& r) { a = r; }).on_success([&](expected<float>&) { a = 91.f; });
-
-    EXPECT_THAT(a, Eq(612.1f));
-}
-
-TEST_F(expected_test, OnErrorWhenHavingAnErrorWithoutResult)
-{
-    expected<int, float> sut{error<float>(112.1f)};
-    int a = 0;
-    sut.on_error([&]() { a = 5; }).on_success([&]() { a = 7; });
-
-    EXPECT_THAT(a, Eq(5));
-}
-
-TEST_F(expected_test, ConstOnErrorWhenHavingAnErrorWithoutResult)
-{
-    const expected<int, float> sut{error<float>(1112.1f)};
-    int a = 0;
-    sut.on_error([&]() { a = 51; }).on_success([&]() { a = 71; });
-
-    EXPECT_THAT(a, Eq(51));
-}
-
-TEST_F(expected_test, VoidOnErrorWhenHavingAnErrorWithoutResult)
-{
-    expected<float> sut{error<float>(4112.1f)};
-    int a = 0;
-    sut.on_error([&]() { a = 53; }).on_success([&]() { a = 37; });
-
-    EXPECT_THAT(a, Eq(53));
-}
-
-TEST_F(expected_test, VoidConstOnErrorWhenHavingAnErrorWithoutResult)
-{
-    const expected<float> sut{error<float>(18112.1f)};
-    int a = 0;
-    sut.on_error([&]() { a = 451; }).on_success([&]() { a = 71; });
-
-    EXPECT_THAT(a, Eq(451));
-}
-
-TEST_F(expected_test, OnSuccessWhenHavingSuccessWithResult)
+TEST_F(expected_test, ValueTypeAndThenWhenHavingSuccessWithResult)
 {
     expected<int, float> sut{success<int>(112)};
     int a = 0;
-    sut.on_error([&](expected<int, float>&) { a = 3; }).on_success([&](expected<int, float>& r) { a = r.get_value(); });
-
-    EXPECT_THAT(a, Eq(112));
-}
-
-TEST_F(expected_test, ConstOnSuccessWhenHavingSuccessWithResult)
-{
-    const expected<int, float> sut{success<int>(1142)};
-    int a = 0;
-    sut.on_error([&](expected<int, float>&) { a = 3; }).on_success([&](expected<int, float>& r) { a = r.get_value(); });
-
-    EXPECT_THAT(a, Eq(1142));
-}
-
-TEST_F(expected_test, ValueTypeOnSuccessWhenHavingSuccessWithResult)
-{
-    expected<int, float> sut{success<int>(112)};
-    int a = 0;
-    sut.on_error([&](expected<int, float>&) { a = 3; }).on_success([&](int& r) { a = r; });
+    sut.and_then([&](int& r) { a = r; }).or_else([&](float&) { a = 3; });
 
     EXPECT_THAT(a, Eq(112));
 }
@@ -312,65 +214,9 @@ TEST_F(expected_test, ValueTypeConstOnSuccessWhenHavingSuccessWithResult)
 {
     const expected<int, float> sut{success<int>(1142)};
     int a = 0;
-    sut.on_error([&](expected<int, float>&) { a = 3; }).on_success([&](int& r) { a = r; });
+    sut.and_then([&](int& r) { a = r; }).or_else([&](float&) { a = 3; });
 
     EXPECT_THAT(a, Eq(1142));
-}
-
-TEST_F(expected_test, VoidOnSuccessWhenHavingSuccessWithResult)
-{
-    expected<float> sut{success<>()};
-    int a = 0;
-    sut.on_error([&](expected<float>&) { a = 3; }).on_success([&](expected<float>& r) { a = (r.has_error()) ? 1 : 2; });
-
-    EXPECT_THAT(a, Eq(2));
-}
-
-TEST_F(expected_test, VoidConstOnSuccessWhenHavingSuccessWithResult)
-{
-    const expected<float> sut{success<>()};
-    int a = 0;
-    sut.on_error([&](expected<float>&) { a = 3; }).on_success([&](expected<float>& r) {
-        a = (r.has_error()) ? 13 : 23;
-    });
-
-    EXPECT_THAT(a, Eq(23));
-}
-
-TEST_F(expected_test, OnSuccessWhenHavingSuccessWithoutResult)
-{
-    expected<int, float> sut{success<int>(112)};
-    int a = 0;
-    sut.on_error([&]() { a = 3; }).on_success([&]() { a = 55; });
-
-    EXPECT_THAT(a, Eq(55));
-}
-
-TEST_F(expected_test, ConstOnSuccessWhenHavingSuccessWithoutResult)
-{
-    const expected<int, float> sut{success<int>(1125)};
-    int a = 0;
-    sut.on_error([&]() { a = 3; }).on_success([&]() { a = 555; });
-
-    EXPECT_THAT(a, Eq(555));
-}
-
-TEST_F(expected_test, VoidOnSuccessWhenHavingSuccessWithoutResult)
-{
-    expected<float> sut{success<>()};
-    int a = 0;
-    sut.on_error([&]() { a = 3; }).on_success([&]() { a = 855; });
-
-    EXPECT_THAT(a, Eq(855));
-}
-
-TEST_F(expected_test, VoidConstOnSuccessWhenHavingSuccessWithoutResult)
-{
-    const expected<float> sut{success<>()};
-    int a = 0;
-    sut.on_error([&]() { a = 3; }).on_success([&]() { a = 8553; });
-
-    EXPECT_THAT(a, Eq(8553));
 }
 
 TEST_F(expected_test, ConvertNonVoidSuccessResultToVoidResult)
@@ -386,4 +232,19 @@ TEST_F(expected_test, ConvertNonVoidErrorResultToVoidResult)
     expected<float> sut2 = sut;
     EXPECT_THAT(sut2.has_error(), Eq(true));
     EXPECT_THAT(sut2.get_error(), Eq(1.23f));
+}
+
+TEST_F(expected_test, ExpectedWithValueConvertsToOptionalWithValue)
+{
+    expected<int, float> sut{success<int>(4711)};
+    optional<int> value = sut.to_optional();
+    ASSERT_THAT(value.has_value(), Eq(true));
+    EXPECT_THAT(*value, Eq(4711));
+}
+
+TEST_F(expected_test, ExpectedWithErrorConvertsToNullopt)
+{
+    expected<int, float> sut{error<float>(47.11f)};
+    optional<int> value = sut.to_optional();
+    ASSERT_THAT(value.has_value(), Eq(false));
 }

--- a/iceoryx_utils/test/moduletests/test_cxx_optional.cpp
+++ b/iceoryx_utils/test/moduletests/test_cxx_optional.cpp
@@ -12,8 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "test.hpp"
 #include "iceoryx_utils/cxx/optional.hpp"
+#include "test.hpp"
 
 
 using namespace ::testing;
@@ -464,4 +464,10 @@ TEST_F(Optional_test, OrElseWhenNotContainingValue)
 
     sut.or_else([&] { value = 42; });
     EXPECT_THAT(value, Eq(42));
+}
+
+TEST_F(Optional_test, ReturningNulloptWithoutConstruction)
+{
+    auto val = []() -> iox::cxx::optional<int> { return iox::cxx::nullopt; }();
+    EXPECT_THAT(val.has_value(), Eq(false));
 }

